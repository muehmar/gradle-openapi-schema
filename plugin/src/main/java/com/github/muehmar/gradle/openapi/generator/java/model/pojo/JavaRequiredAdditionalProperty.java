package com.github.muehmar.gradle.openapi.generator.java.model.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.model.type.JavaAnyType.javaAnyType;
import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import lombok.Value;

@Value
public class JavaRequiredAdditionalProperty {
  JavaName name;
  JavaType javaType;

  public static JavaRequiredAdditionalProperty fromNameAndType(Name name, JavaType javaType) {
    return new JavaRequiredAdditionalProperty(JavaName.fromName(name), javaType);
  }

  public String getDescription() {
    return String.format("Additional Property '%s'", name);
  }

  public boolean isAnyType() {
    return javaType.equals(javaAnyType());
  }

  public boolean isNotAnyType() {
    return not(isAnyType());
  }
}
