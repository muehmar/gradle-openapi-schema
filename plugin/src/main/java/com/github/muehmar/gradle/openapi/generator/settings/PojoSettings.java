package com.github.muehmar.gradle.openapi.generator.settings;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import com.github.muehmar.gradle.openapi.task.TaskIdentifier;
import com.github.muehmar.gradle.openapi.warnings.Warning;
import com.github.muehmar.gradle.openapi.warnings.WarningsContext;
import io.github.muehmar.pojobuilder.annotations.FieldBuilder;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.io.Serializable;
import java.util.List;
import lombok.Value;
import lombok.With;

@Value
@With
@PojoBuilder
public class PojoSettings implements Serializable {
  JsonSupport jsonSupport;
  String packageName;
  String suffix;
  StagedBuilderSettings stagedBuilder;
  String builderMethodPrefix;
  boolean enableValidation;
  boolean nonStrictOneOfValidation;

  ValidationApi validationApi;
  List<ClassTypeMapping> classTypeMappings;
  List<FormatTypeMapping> formatTypeMappings;
  EnumDescriptionSettings enumDescriptionSettings;
  GetterSuffixes getterSuffixes;
  ValidationMethods validationMethods;
  List<String> excludeSchemas;
  PojoNameMappings pojoNameMappings;

  TaskIdentifier taskIdentifier;

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableValidation() {
    return enableValidation;
  }

  public boolean isNonStrictOneOfValidation() {
    return nonStrictOneOfValidation;
  }

  public List<ClassTypeMapping> getClassTypeMappingsList() {
    return classTypeMappings;
  }

  public List<FormatTypeMapping> getFormatTypeMappingsList() {
    return formatTypeMappings;
  }

  public List<String> getExcludeSchemasList() {
    return excludeSchemas;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public ExcludedSchemas getExcludedSchemas() {
    return ExcludedSchemas.fromExcludedPojoNames(
        PList.fromIter(excludeSchemas).map(Name::ofString));
  }

  public boolean isEnableStagedBuilder() {
    return stagedBuilder.isEnabled();
  }

  public boolean isDisableStagedBuilder() {
    return !isEnableStagedBuilder();
  }

  public StagedBuilderSettings getStagedBuilder() {
    return stagedBuilder;
  }

  @FieldBuilder(fieldName = "classTypeMappings")
  public static List<ClassTypeMapping> classTypeMappings(
      PList<ClassTypeMapping> classTypeMappings) {
    return classTypeMappings.toArrayList();
  }

  @FieldBuilder(fieldName = "formatTypeMappings")
  public static List<FormatTypeMapping> formatTypeMappings(
      PList<FormatTypeMapping> formatTypeMappings) {
    return formatTypeMappings.toArrayList();
  }

  public TypeMappings getTypeMappings() {
    return new TypeMappings(PList.fromIter(classTypeMappings), PList.fromIter(formatTypeMappings));
  }

  public PojoNameMapping pojoNameMapping() {
    return PList.fromIter(pojoNameMappings.getConstantNameMappings())
        .map(
            constantNameMapping ->
                PojoNameMapping.replaceConstant(
                    constantNameMapping.getConstant(), constantNameMapping.getReplacement()))
        .foldLeft(PojoNameMapping.noMapping(), PojoNameMapping::andThen);
  }

  public void validate() {
    classTypeMappings.forEach(
        classTypeMapping -> {
          if (not(classTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(classTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });

    formatTypeMappings.forEach(
        formatTypeMapping -> {
          if (not(formatTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(formatTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });
  }
}
