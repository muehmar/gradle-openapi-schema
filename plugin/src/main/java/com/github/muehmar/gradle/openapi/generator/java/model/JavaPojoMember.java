package com.github.muehmar.gradle.openapi.generator.java.model;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumGenerator.EnumContent;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsNullFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsPresentFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaEnumType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.PojoMember;
import com.github.muehmar.gradle.openapi.generator.model.type.EnumType;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixes;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import lombok.With;

/**
 * This member corresponds to a property in the specification. To support all possible states af a
 * member, a single {@link JavaPojoMember} may result in more than one {@link TechnicalPojoMember}
 * in a pojo.
 */
@EqualsAndHashCode
@ToString
@PojoBuilder
@With
public class JavaPojoMember {
  private final JavaName name;
  private final String description;
  private final JavaType javaType;
  private final Necessity necessity;
  private final Nullability nullability;
  private final MemberType type;

  private static final String TRISTATE_TO_PROPERTY =
      "onValue(val -> val).onNull(() -> null).onAbsent(() -> null)";
  private static final String TRISTATE_TO_ISNULL_FLAG =
      "onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false)";

  JavaPojoMember(
      JavaName name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability,
      MemberType type) {
    this.javaType = javaType;
    this.name = name;
    this.description = description;
    this.necessity = necessity;
    this.nullability = nullability;
    this.type = type;
  }

  public static JavaPojoMember wrap(PojoMember pojoMember, TypeMappings typeMappings) {
    final JavaType javaType = JavaType.wrap(pojoMember.getType(), typeMappings);
    return new JavaPojoMember(
        JavaName.fromName(pojoMember.getName()),
        pojoMember.getDescription(),
        javaType,
        pojoMember.getNecessity(),
        pojoMember.getNullability(),
        MemberType.OBJECT_MEMBER);
  }

  public JavaName getName() {
    return name;
  }

  public Nullability getNullability() {
    return nullability;
  }

  public Necessity getNecessity() {
    return necessity;
  }

  public String getDescription() {
    return description;
  }

  public JavaType getJavaType() {
    return javaType;
  }

  public MemberType getType() {
    return type;
  }

  public JavaPojoMember asObjectMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.OBJECT_MEMBER)
        .andAllOptionals()
        .build();
  }

  public JavaPojoMember asAllOfMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.ALL_OF_MEMBER)
        .andAllOptionals()
        .build();
  }

  public JavaPojoMember asOneOfMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.ONE_OF_MEMBER)
        .andAllOptionals()
        .build();
  }

  public JavaPojoMember asAnyOfMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.ANY_OF_MEMBER)
        .andAllOptionals()
        .build();
  }

  public boolean isOptional() {
    return necessity.isOptional();
  }

  public boolean isRequired() {
    return necessity.isRequired();
  }

  public boolean isNullable() {
    return nullability.isNullable();
  }

  public boolean isNotNullable() {
    return nullability.isNotNullable();
  }

  public boolean isRequiredAndNullable() {
    return isRequired() && isNullable();
  }

  public boolean isRequiredAndNotNullable() {
    return isRequired() && isNotNullable();
  }

  public boolean isOptionalAndNullable() {
    return isOptional() && isNullable();
  }

  public boolean isOptionalAndNotNullable() {
    return isOptional() && isNotNullable();
  }

  public JavaName getWitherName() {
    return prefixedMethodName("with");
  }

  public JavaName getIsPresentFlagName() {
    return IsPresentFlagName.fromName(name).getName();
  }

  public JavaName getIsNullFlagName() {
    return IsNullFlagName.fromName(name).getName();
  }

  public JavaName getGetterName() {
    return prefixedMethodName("get");
  }

  public JavaName getValidationGetterName(PojoSettings settings) {
    return JavaName.fromString(getGetterName().asString())
        .append(settings.getValidationMethods().getGetterSuffix());
  }

  public JavaName getGetterNameWithSuffix(PojoSettings settings) {
    return JavaName.fromString(getGetterName().asString()).append(determineSuffix(settings));
  }

  private String determineSuffix(PojoSettings settings) {
    final GetterSuffixes getterSuffixes = settings.getGetterSuffixes();
    if (isRequiredAndNotNullable()) {
      return getterSuffixes.getRequiredSuffix();
    } else if (isRequiredAndNullable()) {
      return getterSuffixes.getRequiredNullableSuffix();
    } else if (isOptionalAndNotNullable()) {
      return getterSuffixes.getOptionalSuffix();
    } else {
      return getterSuffixes.getOptionalNullableSuffix();
    }
  }

  public JavaName prefixedMethodName(String prefix) {
    return name.prefixedMethodName(prefix);
  }

  public PList<TechnicalPojoMember> getTechnicalMembers() {
    final TechnicalPojoMember technicalPojoMember = TechnicalPojoMember.wrapPojoMember(this);
    if (isRequiredAndNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isPresentFlagMember(getIsPresentFlagName()));
    } else if (isOptionalAndNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isNullFlagMember(getIsNullFlagName()));
    } else {
      return PList.single(technicalPojoMember);
    }
  }

  public String tristateToProperty() {
    return TRISTATE_TO_PROPERTY;
  }

  public String tristateToIsNullFlag() {
    return TRISTATE_TO_ISNULL_FLAG;
  }

  /** Creates {@link EnumContent} for this member in case its type is an {@link EnumType}. */
  public Optional<EnumContent> asEnumContent() {
    final Function<JavaEnumType, Optional<EnumContent>> toEnumPojo =
        enumType ->
            Optional.of(
                EnumContentBuilder.create()
                    .className(JavaName.fromName(enumType.getQualifiedClassName().getClassName()))
                    .description(getDescription())
                    .members(enumType.getMembers())
                    .build());
    return javaType.fold(
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        toEnumPojo,
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty());
  }

  /**
   * In case this member is an enum type, it converts the classname of the enum so that it's
   * referenced via outer-class {@code javaPojoName}.
   */
  public JavaPojoMember asInnerEnumOf(JavaName javaPojoName) {
    final JavaType newType =
        javaType.fold(
            arrayType -> arrayType,
            booleanType -> booleanType,
            enumType -> enumType.asInnerClassOf(javaPojoName),
            mapType -> mapType,
            noType -> noType,
            numericType -> numericType,
            integerType -> integerType,
            objectType -> objectType,
            stringType -> stringType);
    return new JavaPojoMember(name, description, newType, necessity, nullability, type);
  }

  public enum MemberType {
    OBJECT_MEMBER(false),
    ALL_OF_MEMBER(true),
    ONE_OF_MEMBER(true),
    ANY_OF_MEMBER(true),
    ARRAY_VALUE(false);

    private final boolean isComposedMember;

    MemberType(boolean isComposedMember) {
      this.isComposedMember = isComposedMember;
    }

    public boolean isComposedMember() {
      return isComposedMember;
    }
  }
}
