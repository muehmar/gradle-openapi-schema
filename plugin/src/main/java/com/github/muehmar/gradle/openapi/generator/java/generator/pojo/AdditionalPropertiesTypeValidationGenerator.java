package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.ValidationGenerator.assertTrue;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties.additionalPropertiesName;
import static io.github.muehmar.codegenerator.Generator.constant;

import com.github.muehmar.gradle.openapi.generator.java.generator.shared.SettingsFunctions;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.MethodGenBuilder;

public class AdditionalPropertiesTypeValidationGenerator {
  private AdditionalPropertiesTypeValidationGenerator() {}

  public static Generator<JavaObjectPojo, PojoSettings>
      additionalPropertiesTypeValidationGenerator() {
    return annotatedCorrectTypeMethod()
        .filter(JavaAdditionalProperties::isAllowed)
        .filter(JavaAdditionalProperties::isNotValueAnyType)
        .contraMap(JavaObjectPojo::getAdditionalProperties);
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> annotatedCorrectTypeMethod() {
    return JacksonAnnotationGenerator.<JavaAdditionalProperties>jsonIgnore()
        .append(
            assertTrue(
                props ->
                    String.format(
                        "Not all additional properties are instances of %s",
                        props.getType().getQualifiedClassName().getClassName())))
        .append(correctTypeMethod());
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> correctTypeMethod() {
    return MethodGenBuilder.<JavaAdditionalProperties, PojoSettings>create()
        .modifiers(SettingsFunctions::validationMethodModifiers)
        .noGenericTypes()
        .returnType("boolean")
        .methodName("isAllAdditionalPropertiesHaveCorrectType")
        .noArguments()
        .content(
            constant(
                String.format(
                    "return getAdditionalProperties().size() == %s.size();",
                    additionalPropertiesName())))
        .build();
  }
}
