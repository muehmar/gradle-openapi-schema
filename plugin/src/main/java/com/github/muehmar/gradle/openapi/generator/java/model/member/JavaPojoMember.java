package com.github.muehmar.gradle.openapi.generator.java.model.member;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumGenerator.EnumContent;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsNotNullFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsNullFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsPresentFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaPojoName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.PropertyInfoName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaEnumType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.PojoMember;
import com.github.muehmar.gradle.openapi.generator.model.type.EnumType;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixes;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import lombok.With;

/**
 * This member corresponds to a property in the specification. To support all possible states af a
 * member, a single {@link JavaPojoMember} may result in more than one {@link TechnicalPojoMember}
 * in a pojo.
 */
@EqualsAndHashCode
@ToString
@PojoBuilder
@With
public class JavaPojoMember {
  private final JavaPojoName pojoName;
  private final JavaName name;
  private final String description;
  private final JavaType javaType;
  private final Necessity necessity;
  private final Nullability nullability;
  private final MemberType type;

  private static final String TRISTATE_TO_PROPERTY =
      "onValue(val -> val).onNull(() -> null).onAbsent(() -> null)";
  private static final String TRISTATE_TO_ISNULL_FLAG =
      "onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false)";

  JavaPojoMember(
      JavaPojoName pojoName,
      JavaName name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability,
      MemberType type) {
    this.pojoName = pojoName;
    this.javaType = javaType;
    this.name = name;
    this.description = description;
    this.necessity = necessity;
    this.nullability = nullability;
    this.type = type;
  }

  public static JavaPojoMember wrap(
      PojoMember pojoMember, JavaPojoName pojoName, TypeMappings typeMappings) {
    final JavaType javaType = JavaType.wrap(pojoMember.getType(), typeMappings);
    return new JavaPojoMember(
        pojoName,
        JavaName.fromName(pojoMember.getName()),
        pojoMember.getDescription(),
        javaType,
        pojoMember.getNecessity(),
        pojoMember.getNullability(),
        MemberType.OBJECT_MEMBER);
  }

  public PropertyInfoName getPropertyInfoName() {
    return PropertyInfoName.fromPojoNameAndMemberName(pojoName, name);
  }

  public JavaName getName() {
    return name;
  }

  public Nullability getNullability() {
    return nullability;
  }

  public Necessity getNecessity() {
    return necessity;
  }

  public String getDescription() {
    return description;
  }

  public JavaType getJavaType() {
    return javaType;
  }

  public MemberType getType() {
    return type;
  }

  public boolean isTechnicallyEquals(JavaPojoMember other) {
    return this.getMemberKey().equals(other.getMemberKey());
  }

  public MemberKey getMemberKey() {
    return new MemberKey(name, javaType);
  }

  public JavaPojoMember asAllOfMember() {
    return withType(MemberType.ALL_OF_MEMBER);
  }

  public JavaPojoMember asOneOfMember() {
    return withType(MemberType.ONE_OF_MEMBER);
  }

  public JavaPojoMember asAnyOfMember() {
    return withType(MemberType.ANY_OF_MEMBER);
  }

  public boolean isOptional() {
    return necessity.isOptional();
  }

  public boolean isRequired() {
    return necessity.isRequired();
  }

  public boolean isNullable() {
    return nullability.isNullable();
  }

  public boolean isNotNullable() {
    return nullability.isNotNullable();
  }

  public boolean isRequiredAndNullable() {
    return isRequired() && isNullable();
  }

  public boolean isRequiredAndNotNullable() {
    return isRequired() && isNotNullable();
  }

  public boolean isOptionalAndNullable() {
    return isOptional() && isNullable();
  }

  public boolean isOptionalAndNotNullable() {
    return isOptional() && isNotNullable();
  }

  public JavaName getWitherName() {
    return prefixedMethodName("with");
  }

  public JavaName getIsPresentFlagName() {
    return IsPresentFlagName.fromName(name).getName();
  }

  public JavaName getIsNullFlagName() {
    return IsNullFlagName.fromName(name).getName();
  }

  public JavaName getIsNotNullFlagName() {
    return IsNotNullFlagName.fromName(name).getName();
  }

  public JavaName getGetterName() {
    return prefixedMethodName("get");
  }

  public JavaName getValidationGetterName(PojoSettings settings) {
    return JavaName.fromString(getGetterName().asString())
        .append(settings.getValidationMethods().getGetterSuffix());
  }

  public JavaName getGetterNameWithSuffix(PojoSettings settings) {
    return JavaName.fromString(getGetterName().asString()).append(determineSuffix(settings));
  }

  private String determineSuffix(PojoSettings settings) {
    final GetterSuffixes getterSuffixes = settings.getGetterSuffixes();
    if (isRequiredAndNotNullable()) {
      return getterSuffixes.getRequiredSuffix();
    } else if (isRequiredAndNullable()) {
      return getterSuffixes.getRequiredNullableSuffix();
    } else if (isOptionalAndNotNullable()) {
      return getterSuffixes.getOptionalSuffix();
    } else {
      return getterSuffixes.getOptionalNullableSuffix();
    }
  }

  public JavaName prefixedMethodName(String prefix) {
    return name.prefixedMethodName(prefix);
  }

  public PList<TechnicalPojoMember> getTechnicalMembers() {
    final TechnicalPojoMember technicalPojoMember = TechnicalPojoMember.wrapPojoMember(this);
    if (isRequiredAndNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isPresentFlagMember(getIsPresentFlagName()));
    } else if (isOptionalAndNotNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isNotNullFlagMember(getIsNotNullFlagName()));
    } else if (isOptionalAndNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isNullFlagMember(getIsNullFlagName()));
    } else {
      return PList.single(technicalPojoMember);
    }
  }

  public String tristateToProperty() {
    return TRISTATE_TO_PROPERTY;
  }

  public String tristateToIsNullFlag() {
    return TRISTATE_TO_ISNULL_FLAG;
  }

  public Optional<JavaPojoMember> mergeToLeastRestrictive(JavaPojoMember other) {
    if (this.getMemberKey().equals(other.getMemberKey())) {
      final Nullability leastRestrictiveNullability =
          Nullability.leastRestrictive(this.getNullability(), other.getNullability());
      final Necessity leastRestrictiveNecessity =
          Necessity.leastRestrictive(this.getNecessity(), other.getNecessity());
      return Optional.of(
          this.withNullability(leastRestrictiveNullability)
              .withNecessity(leastRestrictiveNecessity));
    } else {
      return Optional.empty();
    }
  }

  public Optional<JavaPojoMember> mergeToMostRestrictive(JavaPojoMember other) {
    if (this.getMemberKey().equals(other.getMemberKey())) {
      final Nullability mostRestrictiveNullability =
          Nullability.mostRestrictive(this.getNullability(), other.getNullability());
      final Necessity mostRestrictiveNecessity =
          Necessity.mostRestrictive(this.getNecessity(), other.getNecessity());
      return Optional.of(
          this.withNullability(mostRestrictiveNullability).withNecessity(mostRestrictiveNecessity));
    } else {
      return Optional.empty();
    }
  }

  /** Creates {@link EnumContent} for this member in case its type is an {@link EnumType}. */
  public Optional<EnumContent> asEnumContent() {
    final Function<JavaEnumType, Optional<EnumContent>> toEnumPojo =
        enumType ->
            Optional.of(
                EnumContentBuilder.create()
                    .className(JavaName.fromName(enumType.getQualifiedClassName().getClassName()))
                    .description(getDescription())
                    .members(enumType.getMembers())
                    .build());
    return javaType.fold(
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        toEnumPojo,
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty());
  }

  /**
   * In case this member is an enum type, it converts the classname of the enum so that it's
   * referenced via outer-class {@code javaPojoName}.
   */
  public JavaPojoMember asInnerEnumOf(JavaName javaPojoName) {
    final JavaType newType =
        javaType.fold(
            arrayType -> arrayType,
            booleanType -> booleanType,
            enumType -> enumType.asInnerClassOf(javaPojoName),
            mapType -> mapType,
            noType -> noType,
            numericType -> numericType,
            integerType -> integerType,
            objectType -> objectType,
            stringType -> stringType);
    return new JavaPojoMember(pojoName, name, description, newType, necessity, nullability, type);
  }

  public enum MemberType {
    OBJECT_MEMBER,
    ADDITIONAL_PROPERTY_MEMBER,
    ALL_OF_MEMBER,
    ONE_OF_MEMBER,
    ANY_OF_MEMBER,
    ARRAY_VALUE
  }
}
