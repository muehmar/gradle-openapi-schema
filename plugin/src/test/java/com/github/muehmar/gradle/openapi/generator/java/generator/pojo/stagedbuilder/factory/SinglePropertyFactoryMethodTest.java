package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.stagedbuilder.factory;

import static ch.bluecare.commons.data.PList.single;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static org.junit.jupiter.api.Assertions.assertEquals;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.member.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.member.JavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.java.model.member.TestJavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaRequiredAdditionalProperty;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaTypes;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import java.util.stream.Stream;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

@SnapshotTest
class SinglePropertyFactoryMethodTest {
  private Expect expect;

  @ParameterizedTest
  @MethodSource("members")
  @SnapshotName("members")
  void singlePropertyFactoryMethod_when_members_then_matchSnapshot(PList<JavaPojoMember> members) {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        SinglePropertyFactoryMethod.singlePropertyFactoryMethod();

    final Writer writer =
        generator.generate(
            JavaPojos.objectPojo(members),
            TestPojoSettings.defaultTestSettings(),
            Writer.javaWriter());

    final String scenario = members.map(JavaPojoMember::getName).mkString(", ");

    expect.scenario(scenario).toMatchSnapshot(writerSnapshot(writer));
  }

  static Stream<Arguments> members() {
    return Stream.of(
            single(TestJavaPojoMembers.requiredBirthdate()),
            single(TestJavaPojoMembers.optionalString()),
            single(TestJavaPojoMembers.optionalNullableString()),
            PList.of(TestJavaPojoMembers.requiredString(), TestJavaPojoMembers.optionalBirthdate()))
        .map(Arguments::of);
  }

  @ParameterizedTest
  @MethodSource("noSinglePropertyObjectPojos")
  void singlePropertyFactoryMethod_when_noSingleMemberPojos_then_noOutput(JavaObjectPojo pojo) {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        SinglePropertyFactoryMethod.singlePropertyFactoryMethod();

    final Writer writer =
        generator.generate(pojo, TestPojoSettings.defaultTestSettings(), Writer.javaWriter());

    assertEquals("", writer.asString());
  }

  public static Stream<Arguments> noSinglePropertyObjectPojos() {
    final JavaPojoMembers singleMember =
        JavaPojoMembers.fromMembers(single(TestJavaPojoMembers.requiredBirthdate()));
    return Stream.of(
            JavaPojos.anyOfPojo(JavaPojos.sampleObjectPojo1(), JavaPojos.sampleObjectPojo2())
                .withMembers(singleMember),
            JavaPojos.oneOfPojo(JavaPojos.sampleObjectPojo1(), JavaPojos.sampleObjectPojo2())
                .withMembers(singleMember),
            JavaPojos.allOfPojo(JavaPojos.sampleObjectPojo1(), JavaPojos.sampleObjectPojo2())
                .withMembers(singleMember),
            JavaPojos.sampleObjectPojo1(),
            JavaPojos.objectPojo(singleMember.asList())
                .withRequiredAdditionalProperties(
                    PList.single(
                        JavaRequiredAdditionalProperty.fromNameAndType(
                            Name.ofString("prop"), JavaTypes.stringType()))))
        .map(Arguments::of);
  }
}
