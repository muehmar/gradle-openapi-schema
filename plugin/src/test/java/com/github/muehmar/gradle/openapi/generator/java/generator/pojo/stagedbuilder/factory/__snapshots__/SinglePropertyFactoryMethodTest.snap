members[birthdate]=[
java.time.LocalDate

/**
 * Convenience factory method for DTO's with a single property. Instantiates
 * directly the DTO with the given property. This method is only generated if the
 * DTO has exactly one property.
 */
public static ObjectPojo1Dto fromBirthdate(LocalDate birthdate) {
  return fullBuilder().setBirthdate(birthdate).build();
}

]


members[optionalNullableStringVal]=[
com.github.muehmar.openapi.util.Tristate

/**
 * Convenience factory method for DTO's with a single property. Instantiates
 * directly the DTO with the given property. This method is only generated if the
 * DTO has exactly one property.
 */
public static ObjectPojo1Dto fromOptionalNullableStringVal(String optionalNullableStringVal) {
  return fullBuilder().setOptionalNullableStringVal(optionalNullableStringVal).build();
}

/**
 * Convenience factory method for DTO's with a single property. Instantiates
 * directly the DTO with the given property. This method is only generated if the
 * DTO has exactly one property.
 */
public static ObjectPojo1Dto fromOptionalNullableStringVal(Tristate<String> optionalNullableStringVal) {
  return fullBuilder().setOptionalNullableStringVal(optionalNullableStringVal).build();
}
]


members[optionalStringVal]=[
java.util.Optional

/**
 * Convenience factory method for DTO's with a single property. Instantiates
 * directly the DTO with the given property. This method is only generated if the
 * DTO has exactly one property.
 */
public static ObjectPojo1Dto fromOptionalStringVal(String optionalStringVal) {
  return fullBuilder().setOptionalStringVal(optionalStringVal).build();
}

/**
 * Convenience factory method for DTO's with a single property. Instantiates
 * directly the DTO with the given property. This method is only generated if the
 * DTO has exactly one property.
 */
public static ObjectPojo1Dto fromOptionalStringVal(Optional<String> optionalStringVal) {
  return fullBuilder().setOptionalStringVal(optionalStringVal).build();
}

]


members[stringVal, birthdate]=[
.
.

]